### PR TITLE
Fix the page fault handling performance regression

### PR DESCRIPTION
**Pull request by junyang-zh**
#1158 introduces some performance regression on the lmbench page fault benchmark.

On the `INTEL(R) XEON(R) PLATINUM 8575C` CI machine, for the current main branch the result is:

```
[
  {
    "name": "Average page fault latency on Linux",
    "unit": "µs",
    "value": "0.1110",
    "extra": "linux_result"
  },
  {
    "name": "Average page fault latency on Asterinas",
    "unit": "µs",
    "value": "0.6302",
    "extra": "aster_result"
  }
]
```

After the first commit:

```
[
  {
    "name": "Average page fault latency on Linux",
    "unit": "µs",
    "value": "0.1116",
    "extra": "linux_result"
  },
  {
    "name": "Average page fault latency on Asterinas",
    "unit": "µs",
    "value": "0.5072",
    "extra": "aster_result"
  }
]
```

After the second commit:

```
[
  {
    "name": "Average page fault latency on Linux",
    "unit": "µs",
    "value": "0.1127",
    "extra": "linux_result"
  },
  {
    "name": "Average page fault latency on Asterinas",
    "unit": "µs",
    "value": "0.4647",
    "extra": "aster_result"
  }
]
```

After the third:

```
[
  {
    "name": "Average page fault latency on Linux",
    "unit": "µs",
    "value": "0.1118",
    "extra": "linux_result"
  },
  {
    "name": "Average page fault latency on Asterinas",
    "unit": "µs",
    "value": "0.0998",
    "extra": "aster_result"
  }
]
```

We are even better.

### CODE REVIEW

**Comment by tatetian** [comment]
@cchanging Could you review this PR? Thanks!

**Comment by tatetian** [comment]
Is the regression on "the cost of fork+exec+exit" also caused by #1158?

**Comment by junyang-zh** [comment]
> Is the regression on "the cost of fork+exec+exit" also caused by #1158?

Affirmative. After this fix the performance goes better than 1 day earlier too.

```
[
  {
    "name": "Average exec latency on Linux",
    "unit": "µs",
    "value": "198.5000",
    "extra": "linux_result"
  },
  {
    "name": "Average exec latency on Asterinas",
    "unit": "µs",
    "value": "257.6364",
    "extra": "aster_result"
  }
]
```

File: ostd/src/mm/vm_space.rs [hunk 1]
```diff
@@ -389,6 +396,10 @@ impl CursorMut<'_, '_> {
             }
         }
 
+        if !self.need_remote_flush && tlb_prefer_flush_all && self.need_self_flush {
```
Comment (cchanging) [inline ostd/src/mm/vm_space.rs:399]: Good job! The second commit reveals and resolves the previous performance bottleneck! 

This PR looks good to me, with only a small issue: why checking both `!self.need_remote_flush` and `self.need_self_flush` here? From my understanding, These two conditions should be mutually exclusive, right?

Comment (junyang-zh) [reply ostd/src/mm/vm_space.rs:399]: No. Sometimes there can be no flushes at all. The `VmSpace` is not activated anywhere.

Comment (junyang-zh) [reply ostd/src/mm/vm_space.rs]: But this check makes no effect since `issue_tlb_flush` checks it anyway. I've removed it.

Comment (cchanging) [reply ostd/src/mm/vm_space.rs]: I previously thought that if this code were to be executed, the `VmSpace` would at least be activated on one core, so I assumed that `!self.need_remote_flush` would necessarily mean `self.need_self_flush`. However, from the perspective of the API and the constructor, it does seem as you mentioned. I have no issues with the other changes.

**Comment by junyang-zh** [comment]
CI failed due to something similar as #1304, but in the fcntl implementation.

**Comment by tatetian** [comment]
> CI failed due to something similar to #1304, but in the fcntl implementation.

Thanks for the failure diagnosis.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**